File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/events/GiftCardIssued.java
package it.giannibombelli.wsc2026.giftcard.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

public record GiftCardIssued(GiftCardId aggregateId, Money balance) implements GiftCardEvent {
    public GiftCardIssued {
        Require.requireArgument(aggregateId, "giftCardId");
        Require.requireArgument(balance, "balance");
    }
}

File: typescript/src/giftcard/domain/events/giftCardIssued.ts
// Evento emesso quando una gift card viene emessa.

import type { Event } from '@/common/domain/model/event.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { GiftCardId } from '../giftcard/giftCardId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type GiftCardIssued = Event<'GiftCardIssued', GiftCardId> & {
  readonly balance: Money;
};

export function giftCardIssued(aggregateId: GiftCardId, balance: Money): GiftCardIssued {
  requireArgument(aggregateId, 'giftCardId');
  requireArgument(balance, 'balance');
  return { kind: 'GiftCardIssued', aggregateId, balance };
}
